contracts/StakerV2.sol

```diff
@@ -0,0 +1,948 @@
+    NftUserInfo storage staker =  nftUserInfo[_msgSender()];
+
+    // Iterate through all the boosters that the pool supports
+    for(uint256 i = 0; i < pool.boostInfo.length; i++) {
+      booster = boostInfo[pool.boostInfo[i]];
+      if(staker.boosterIds.contains(pool.boostInfo[i])) 
+        if(booster.assetType == AssetType.Tokens && _isToken)
+          _boosted += (_unboosted * booster.multiplier)/10000;
+        else if(booster.assetType == AssetType.Points && !_isToken)
+          _boosted += (_unboosted * booster.multiplier)/10000;
+        else if (booster.assetType == AssetType.Both)
+           _boosted += (_unboosted * booster.multiplier)/10000;
+    }
+  }
+
+  /**
+    Deposit some particular assets to a particular pool on the Staker.
+    @param _token the asset to stake into its corresponding pool.
+    @param _amount the amount of the provided asset to stake.
+  */
+  function deposit(IERC20 _token, uint256 _amount) external nonReentrant {
+
+    PoolInfo storage pool = poolInfo[_token];
+    require(pool.tokenStrength > 0 || pool.pointStrength > 0,
+      "Inactive pool.");
+    UserInfo storage user = userInfo[_token][msg.sender];
+
+    updatePool(_token);
+    updateDeposits(_amount, _token, pool, user, 0);
+
+    pool.token.safeTransferFrom(msg.sender, address(this), _amount);
+    emit Deposit(msg.sender, _token, _amount);
+  }
+
+  /**
+    Withdraw some particular assets from a particular pool on the Staker.
+    @param _token the asset to withdraw from its corresponding staking pool.
+    @param _amount the amount of the provided asset to withdraw.
+  */
+  function withdraw(IERC20 _token, uint256 _amount) external nonReentrant {
+
+    PoolInfo storage pool = poolInfo[_token];
+    UserInfo storage user = userInfo[_token][msg.sender];
+    require(user.amount >= _amount,
+      "Invalid amount.");
+
+    updatePool(_token);
+    updateDeposits(_amount, _token, pool, user, 1);
+
+    pool.token.safeTransfer(msg.sender, _amount);
+    emit Withdraw(msg.sender, _token, _amount);
+  }
+
+  /**
+    Private helper function to check if NFT staker is eligible for a booster.
+    @param _ids ids of NFTs required for a booster.
+    @param _contract external contract from which NFTs are required.
+    @param _boosterId the booster for which NFTs are being staked.
+    @return return true if eligible.
+   */
+  function eligible(uint256[] memory _ids, address _contract, uint256 _boosterId) private view returns(bool) {
+
+    BoostInfo memory booster = boostInfo[_boosterId];
+    if(booster.multiplier == 0) // Inactive
+      return false;
+    else if(_contract != booster.contractRequired) // Different contract
+      return false;
+    else if(_ids.length < booster.amountRequired) // Insufficient amount
+      return false;
+    else if(booster.groupRequired != 0)
+      for(uint256 i = 0; i < _ids.length; i++)
+        if(_ids[i] >> 128 != booster.groupRequired) // Wrong group item
```

Comment:
should we also have a check here to make sure that `_ids[i] >> 128 != 0`  or am i being paranoid?